src/Storages/MergeTree/IMergeTreeDataPart.cpp

```diff
@@ -358,6 +358,12 @@ void IMergeTreeDataPart::setIndex(Columns index_)
     index_loaded = true;
 }
 
+void IMergeTreeDataPart::unloadIndex()
+{
+    std::scoped_lock lock(index_mutex);
+    index.clear();
+    index_loaded = false;
```

Comment:
What will happen if some queries are still using the index?

Reply:
That's an excellent question :thinking: 

I took a look and since read queries that use `getIndex` get a reference to underlying `index` as it is, inflight queries would be affected by the `SYSTEM UNLOAD PRIMARY KEY` that clears it.

There are several solutions to this, but the one that I found simpler to implement, to reason about and doesn't need expensive copies is to promote `index` to a `shared_ptr`.  [Here](https://github.com/ClickHouse/ClickHouse/pull/62738/commits/39deb3ee7f74d6249f1e57238d14489fde6cb019)'s the commit with it.

I saw that the `IColumn` interface is using a copy-on-write shared pointer, so my 2nd option was to create a new `std::vector` where I assign every column from the original `ìndex`. However, that implies increasing the refcount for every single column, so IMO it's cheaper and easier to understand doing that in the parent directly.

Reply:
Yes, using shared_ptr is good solution, I would do exactly the same